doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -926,14 +909,257 @@ the conversion in hypotheses :n:`{+ @ident}`.
+----------------------------------------------------------------------
+
+Coq provides several reduction strategies that tactics can invoke and three
+different algorithms to check the convertibility of types:
+
+- :tacn:`lazy` lazily compares applicative terms.
+- :tacn:`vm_compute` is a brute-force but efficient algorithm that
+  first normalizes the
+  terms before comparing them. It is based on a
+  bytecode representation of terms similar to the bytecode
+  representation used in the ZINC virtual machine :cite:`Leroy90`. It is
+  especially useful for intensive computation of algebraic values, such
+  as numbers, and for reflection-based tactics.
+- :tacn:`native_compute` uses the same algorithm as :tacn:`vm_compute`,
+  but it's based on converting the Coq code to OCaml.
+
+The commands to fine-tune the reduction strategies and the lazy conversion
+algorithm are described first.
+
+.. cmd:: Opaque {+ @reference }
+
+   Marks the specified constants so tactics won't unfold them with δ-conversion.
+   :gdef:`Unfolding <unfolding>` means to replace a constant by its definition.
+   "Constants" are items defined by :cmd:`Definition` or :cmd:`Let` (with an
+   explicit body), or by a command associated with a definition such as
+   :cmd:`Fixpoint`, etc, or by a proof ended by :cmd:`Defined`.
+
+   .. todo PR: can we be more explicit about the `Fixpoint`, etc.
+
+      Should we have a defined term "unfoldable constants"?  It seems
+      there are constants that _potentially_ may be unfolded if they're
+      not marked opaque, so "unfoldable constants" seems a bit misleading
+
+      Also, we should describe whether/which constants default to transparent
+      or opaque.
+
+   This command accepts the :attr:`global` attribute.  By default, the scope
+   of :cmd:`Opaque` is limited to the current section or module.
+
+   :cmd:`Opaque` also affects Coq's conversion algorithm, causing
+   it to delay unfolding the specified constants as much as possible when it
+   has to check that two distinct applied constants are convertible.
+   See Section :ref:`conversion-rules`.
+
+.. cmd:: Transparent {+ @reference }
+
+   The opposite of :cmd:`Opaque`, it marks the specified constants
+   so that tactics may unfold them.  See :cmd:`Opaque` above.
+
+   This command accepts the :attr:`global` attribute.  By default, the scope
+   of :cmd:`Transparent` is limited to the current section or module.
+
+   Note in particular that constants defined by a proofs that end with :cmd:Qed: are
+   not unfoldable and :cmd:`Transparent` has no effect on them. This is to keep
+   with the usual mathematical practice of *proof irrelevance*: what
+   matters in a mathematical development is the sequence of lemma
+   statements, not their actual proofs. This distinguishes lemmas from
+   the usual defined constants, whose actual values are of course
+   relevant in general.
+
+   .. todo PR: Qed proofs are permanently opaque?  Anything else?
```

Comment:
Yes.  Also modules given a type ascription with `:` rather than `<:`.  Not sure if there are other things.

Reply:
Oh, also constants defined by the `abstract` tactic in a proof closed by `Defined` (but not in tactics-in-terms nor in proofs terminated by `Qed`, both of which result in the body of the constant being inlined.

Reply:
Sounds like a couple obscure cases (or bugs?), so I won't mention these.

Reply:
Not bugs, this is on purpose. But definitely edge cases, so it's OK not to mention them as long as you don't make it sound like you're providing an exhaustive list.